noir/noir-repo/tooling/nargo/src/ops/execute.rs

```diff
@@ -8,73 +10,145 @@ use crate::NargoError;
-            }
-            ACVMStatus::Failure(error) => {
-                let call_stack = match &error {
-                    OpcodeResolutionError::UnsatisfiedConstrain {
-                        opcode_location: ErrorLocation::Resolved(opcode_location),
-                    } => Some(vec![*opcode_location]),
-                    OpcodeResolutionError::BrilligFunctionFailed { call_stack, .. } => {
-                        Some(call_stack.clone())
-                    }
-                    _ => None,
-                };
-
-                return Err(NargoError::ExecutionError(match call_stack {
-                    Some(call_stack) => {
-                        // First check whether we have a runtime assertion message that should be resolved on an ACVM failure
-                        // If we do not have a runtime assertion message, we should check whether the circuit has any hardcoded
-                        // messages associated with a specific `OpcodeLocation`.
-                        // Otherwise return the provided opcode resolution error.
-                        if let Some(assert_message) = assert_message {
-                            ExecutionError::AssertionFailed(assert_message.to_owned(), call_stack)
-                        } else if let Some(assert_message) = circuit.get_assert_message(
-                            *call_stack.last().expect("Call stacks should not be empty"),
-                        ) {
-                            ExecutionError::AssertionFailed(assert_message.to_owned(), call_stack)
-                        } else {
-                            ExecutionError::SolvingError(error)
+struct ProgramExecutor<'a, B: BlackBoxFunctionSolver, F: ForeignCallExecutor> {
+    functions: &'a [Circuit],
+    // This gets built as we run through the program looking at each function call
+    witness_stack: WitnessStack,
+
+    blackbox_solver: &'a B,
+
+    foreign_call_executor: &'a mut F,
+}
+
+impl<'a, B: BlackBoxFunctionSolver, F: ForeignCallExecutor> ProgramExecutor<'a, B, F> {
+    fn new(
+        functions: &'a [Circuit],
+        blackbox_solver: &'a B,
+        foreign_call_executor: &'a mut F,
+    ) -> Self {
+        ProgramExecutor {
+            functions,
+            witness_stack: WitnessStack::default(),
+            blackbox_solver,
+            foreign_call_executor,
+        }
+    }
+
+    fn finalize(self) -> WitnessStack {
+        self.witness_stack
+    }
+
+    #[tracing::instrument(level = "trace", skip_all)]
+    fn execute_circuit(
+        &mut self,
+        circuit: &Circuit,
+        initial_witness: WitnessMap,
+    ) -> Result<WitnessMap, NargoError> {
+        let mut acvm = ACVM::new(self.blackbox_solver, &circuit.opcodes, initial_witness);
+
+        // This message should be resolved by a nargo foreign call only when we have an unsatisfied assertion.
+        let mut assert_message: Option<String> = None;
```

Comment:
`assert_failure_message`?

Reply:
This is copied over from the previous code and `assert_messages` is the field on the `Circuit` struct so I think I'll leave it for this PR. If we want to clarify we can do it in a followup I feel.